src/databases.py

```diff
@@ -32,121 +97,60 @@ class Databases(ops.Object):  # pylint: disable=too-few-public-methods
-        if not container.can_connect():
-            logger.info(
-                "pebble client in the Flask container is not ready, defer database-requires-event"
+        self._charm_state = charm_state
+        self._webserver = webserver
+        self._databases = database_requirers
+        for database_requirer in database_requirers.values():
+            self._charm.framework.observe(
+                database_requirer.on.database_created, self._on_database_requires_event
             )
-            event.defer()
-            return
-        plan = container.get_plan()
-        if FLASK_SERVICE_NAME in plan.services:
-            flask_service = plan.services[FLASK_SERVICE_NAME].to_dict()
-            flask_service["environment"].update(self.get_uris())
-            new_layer = ops.pebble.LayerDict(
-                services={
-                    FLASK_SERVICE_NAME: flask_service,
-                }
+            self._charm.framework.observe(
+                self._charm.on[database_requirer.relation_name].relation_broken,
+                self._on_database_requires_event,
             )
-            container.add_layer("flask-app", new_layer, combine=True)
-        container.replan()
 
-    def _setup_database_requirer(self, relation_name: str, database_name: str) -> DatabaseRequires:
-        """Set up a DatabaseRequires instance.
-
-        The DatabaseRequires instance is an interface between the charm and various data providers.
-        It handles those relations and emit events to help us abstract these integrations.
+    def _update_app_and_unit_status(self, status: ops.StatusBase) -> None:
+        """Update the application and unit status.
 
         Args:
-            relation_name: Name of the data relation
-            database_name: Name of the database (can be overwritten by the provider)
-
-        Returns:
-            DatabaseRequires object produced by the data_platform_libs.v0.data_interfaces library
+            status: the desired application and unit status.
         """
-        database_requirer = DatabaseRequires(
-            self._charm,
-            relation_name=relation_name,
-            database_name=database_name,
-        )
-        self._charm.framework.observe(
-            database_requirer.on.database_created, self._on_database_requires_event
-        )
-        self._charm.framework.observe(
-            self._charm.on[relation_name].relation_broken, self._on_database_requires_event
-        )
-        return database_requirer
-
-    def get_uris(self) -> typing.Dict[str, str]:
-        """Compute DatabaseURI and return it.
+        self._charm.unit.status = status
+        if self._charm.unit.is_leader():
+            self._charm.app.status = status
+
+    def _restart_flask(self) -> None:
+        """Restart or start the flask service if not started with the latest configuration."""
+        try:
+            restart_flask(
```

Comment:
Not in this PR but I just saw that restart_flask does not change the unit status if cannot connect to the container or if the secret is not ready. Won't this impact here?